skrf/calibration/deembedding.py

```diff
@@ -908,4 +912,834 @@ def deembed(self, ntwk):
+        #t22r[x:] = 0
+        t22r[x//2:] = 0 # mhuser: strange here but x seems to be wrong by factor 2
+        e002 = np.fft.fft(np.fft.ifftshift(t22r))
+        e002 = e002[1:n+1]
+        
+        # calc e111 and e112
+        e111 = (s22r - e002) / s12r
+        e112 = (s11r - e001) / s21r
+        
+        # calc e01
+        k = 1
+        test = k * np.sqrt(s21r * (1 - e111 * e112))
+        e01 = zeros((n), dtype = np.complex)
+        for i, value in enumerate(test):
+            if(i>0):
+                if(angle(test[i]) - angle(test[i-1]) > 0):
+                    k = -1 * k
+            # mhuser : is it a problem with complex value cast to real here ?
+            e01[i] = k * np.sqrt(s21r[i] * (1 - e111[i] * e112[i]))
+                
+        # calc e10
+        k = 1
+        test = k * np.sqrt(s12r * (1 - e111 * e112))
+        e10 = zeros((n), dtype = np.complex)
+        for i, value in enumerate(test):
+            if(i>0):
+                if(angle(test[i]) - angle(test[i-1]) > 0):
+                    k = -1 * k
+            # mhuser : is it a problem with complex value cast to real here ?
+            e10[i] = k * np.sqrt(s12r[i] * (1 - e111[i] * e112[i]))
+        
+        # S-parameters are setup correctly
+        if not flag_DC and not flag_df:
+            fixture_model_1r = zeros((n, 2, 2), dtype = np.complex)
+            fixture_model_1r[:, 0, 0] = e001
+            fixture_model_1r[:, 1, 0] = e01
+            fixture_model_1r[:, 0, 1] = e01
+            fixture_model_1r[:, 1, 1] = e111
+            
+            fixture_model_2r = zeros((n, 2, 2), dtype = np.complex)
+            fixture_model_2r[:, 1, 1] = e002
+            fixture_model_2r[:, 0, 1] = e10
+            fixture_model_2r[:, 1, 0] = e10
+            fixture_model_2r[:, 0, 0] = e112
+        
+        
+        # S-parameters are not setup correctly
+        else:
+            # todo implement using Network.interpolate to revert to initial freq axis
+            if flag_DC:
+                raise ValueError("TODO : handle DC point already existing.")
+            if flag_df:
+                raise ValueError("TODO : handle interpolation.")
+        
+        # create the S-parameter objects for the errorboxes
+        s_fixture_model_r1  = Network(frequency = s2xthru.frequency, s = fixture_model_1r, z0 = z11x)
+        s_fixture_model_r2  = Network(frequency = s2xthru.frequency, s = fixture_model_2r, z0 = z11x)
+            
+        # renormalize the S-parameter errorboxes to the original reference impedance (assumed to be 50)
+        s_fixture_model_r1.renormalize(50)
+        s_fixture_model_r2.renormalize(50)
+        s_side1 = s_fixture_model_r1
+        s_side2 = s_fixture_model_r2
+        
+        return (s_side1, s_side2)
+    
+class Ieeep370zc2xthru(Deembedding):
+    """
+    IEEEP370Zc2xThru.m creates error boxes from a test fixture 2x thru and the
+    fixture-dut-fixture S-parameters.

```

Comment:
```suggestion
    Creates error boxes from a test fixture 2x thru and the fixture-dut-fixture S-parameters.

```

Reply:
this is also the case for the previous function, isn't it? What is the difference then?

Reply:
- NZC take only fixture-fixture as input and bisect it.
- ZC take fixture-fixture and fixture-dut-fixture as inputs and accound for a possible impedance variation of the tru of boths.